File: frontend/src/api/territory.ts
import { apiGet } from "./client";
import type { TerritoryDoctorsResponse, TerritoryOpportunityResponse } from "../types/api";

export type TerritoryOpportunityFilters = {
  country?: string;
  opportunityLabel?: string;
  page?: number;
  pageSize?: number;
  sortBy?: string;
  sortDir?: "asc" | "desc";
};

export function getTerritoryOpportunities(filters: TerritoryOpportunityFilters = {}) {
  const params = new URLSearchParams();
  if (filters.country) params.set("country", filters.country);
  if (filters.opportunityLabel) params.set("opportunityLabel", filters.opportunityLabel);
  params.set("page", String(filters.page ?? 1));
  params.set("pageSize", String(filters.pageSize ?? 25));
  if (filters.sortBy) params.set("sortBy", filters.sortBy);
  if (filters.sortDir) params.set("sortDir", filters.sortDir);
  return apiGet<TerritoryOpportunityResponse>(`/api/territory/opportunities?${params.toString()}`);
}

export function getTerritoryDoctors(filters: { country: string; territoryName: string; patchName?: string | null }) {
  const params = new URLSearchParams();
  params.set("country", filters.country);
  params.set("territoryName", filters.territoryName);
  if (filters.patchName) params.set("patchName", filters.patchName);
  return apiGet<TerritoryDoctorsResponse>(`/api/territory/doctors?${params.toString()}`);
}

File: backend/app/routers/territory.py
from __future__ import annotations

from typing import Annotated
from typing import Literal

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from backend.app.database import get_session
from backend.app.schemas.territory import TerritoryDoctorsResponse, TerritoryOpportunityResponse
from backend.app.services.territory_service import TerritoryService

router = APIRouter(prefix="/territory", tags=["territory"])


@router.get(
    "/opportunities",
    response_model=TerritoryOpportunityResponse,
    response_model_by_alias=True,
)
def territory_opportunities(
    session: Annotated[Session, Depends(get_session)],
    country: str | None = None,
    opportunity_label: Annotated[str | None, Query(alias="opportunityLabel")] = None,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(alias="pageSize", ge=1, le=100)] = 25,
    sort_by: Annotated[
        Literal["territoryName", "opportunityLabel", "doctorCount", "totalPrescriptionQty"],
        Query(alias="sortBy"),
    ] = "totalPrescriptionQty",
    sort_dir: Annotated[Literal["asc", "desc"], Query(alias="sortDir")] = "desc",
) -> TerritoryOpportunityResponse:
    return TerritoryService(session).opportunities(
        country=country,
        opportunity_label=opportunity_label,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_dir=sort_dir,
    )


@router.get(
    "/doctors",
    response_model=TerritoryDoctorsResponse,
    response_model_by_alias=True,
)
def territory_doctors(
    session: Annotated[Session, Depends(get_session)],
    country: str,
    territory_name: Annotated[str, Query(alias="territoryName")],
    patch_name: Annotated[str | None, Query(alias="patchName")] = None,
) -> TerritoryDoctorsResponse:
    return TerritoryService(session).doctors(
        country=country,
        territory_name=territory_name,
        patch_name=patch_name,
    )
